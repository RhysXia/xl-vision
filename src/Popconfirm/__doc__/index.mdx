# Popconfirm

点击元素，弹出气泡式的确认框。

## 何时使用

目标元素的操作需要用户进一步的确认时，在目标元素附近弹出浮层提示，询问用户。

和`confirm`弹出的全屏居中模态对话框相比，交互形式更轻量。

:::demo 基本用法

基本用法

```jsx
import React from 'react'
import { Popconfirm, Button } from 'xl-vision'

export default () => {
  const message = 'Are you sure?'

  return (
    <div className='container'>
      <div className='top'>
        <Popconfirm placement='top-start' message={message}>
          <Button theme='primary'>tl</Button>
        </Popconfirm>
        <Popconfirm placement='top' message={message}>
          <Button theme='primary'>top</Button>
        </Popconfirm>
        <Popconfirm placement='top-end' message={message}>
          <Button theme='primary'>tr</Button>
        </Popconfirm>
      </div>
      <div className='center'>
        <div className='left'>
          <Popconfirm placement='left-start' message={message}>
            <Button theme='primary'>lt</Button>
          </Popconfirm>
          <Popconfirm placement='left' message={message}>
            <Button theme='primary'>left</Button>
          </Popconfirm>
          <Popconfirm placement='left-end' message={message}>
            <Button theme='primary'>lb</Button>
          </Popconfirm>
        </div>
        <div className='right'>
          <Popconfirm placement='right-start' message={message}>
            <Button theme='primary'>rt</Button>
          </Popconfirm>
          <Popconfirm placement='right' message={message}>
            <Button theme='primary'>right</Button>
          </Popconfirm>
          <Popconfirm placement='right-end' message={message}>
            <Button theme='primary'>rb</Button>
          </Popconfirm>
        </div>
      </div>
      <div className='bottom'>
        <Popconfirm placement='bottom-start' message={message}>
          <Button theme='primary'>bl</Button>
        </Popconfirm>
        <Popconfirm placement='bottom' message={message}>
          <Button theme='primary'>bottom</Button>
        </Popconfirm>
        <Popconfirm placement='bottom-end' message={message}>
          <Button theme='primary'>br</Button>
        </Popconfirm>
      </div>
    </div>
  )
}
```

```scss
.container {
  &::deep {
    .xl-button {
      margin: 10px;
      min-width: 100px;
    }
  }

  .top,
  .bottom {
    margin-left: 100px + 10px;
  }
  .center {
    display: flex;
  }
  .left,
  .right {
    display: flex;
    flex-direction: column;
  }
  .right {
    margin-left: (100px + 10px) * 3;
  }
}
```

:::

:::demo 不同的触发方式

通过`trigger`指定触发方式，默认值是`hover`，如果有复杂需求，可以设置为`custom`，以获得完全的控制能力。

```jsx
import React from 'react'
import { Popconfirm, Button } from 'xl-vision'
export default () => {
  const message = 'Are you sure?'
  const [visible, setVisible] = React.useState(false)

  const onClick = React.useCallback(() => {
    setVisible((prev) => !prev)
  }, [])

  return (
    <Button.Group theme='info'>
      <Popconfirm placement='top' trigger='hover' message={message}>
        <Button>hover</Button>
      </Popconfirm>
      <Popconfirm placement='top' trigger='click' message={message}>
        <Button>click</Button>
      </Popconfirm>
      <Popconfirm placement='top' trigger='focus' message={message}>
        <Button>focus</Button>
      </Popconfirm>
      <Popconfirm placement='top' trigger='contextMenu' message={message}>
        <Button>contextMenu</Button>
      </Popconfirm>
      <Popconfirm placement='top' trigger='custom' visible={visible} message={message}>
        <Button onClick={onClick}>custom(double click)</Button>
      </Popconfirm>
    </Button.Group>
  )
}
```

:::

:::demo 延迟触发

通过`hideDelay`和`showDelay`指定延时触发时间。

```jsx
import React from 'react'
import { Popconfirm, Button } from 'xl-vision'
export default () => {
  const message = 'Are you sure?'

  return (
    <Button.Group theme='secondary'>
      <Popconfirm placement='top' showDelay={2000} message={message}>
        <Button>show after 2s</Button>
      </Popconfirm>
      <Popconfirm placement='top' hideDelay={2000} message={message}>
        <Button>hide after 2s</Button>
      </Popconfirm>
    </Button.Group>
  )
}
```

:::

:::demo 禁用

禁止`popconfirm`展示。

```jsx
import React from 'react'
import { Popconfirm, Button } from 'xl-vision'
export default () => {
  const message = 'Are you sure?'

  return (
    <Popconfirm disabled message={message}>
      <Button theme='primary'>button</Button>
    </Popconfirm>
  )
}
```

:::

## Popconfirm API

> 更多参数请参考<Link to="Popper">**Popper**</Link>属性

| 参数 | 说明 | 类型 | 可选值 | 默认值 | 是否必填 |
| --- | --- | --- | --- | --- | --- |
| clsPrefix | 自定义类名前缀 | string | - | - | false |
| message | 内容 | React.ReactNode | - | - | true |
| footer | 自定义底部按钮 | React.ReactNode | - | - | false |
| confirmText | 确认按钮文字 | string | - | - | false |
| cancelText | 取消按钮文字 | string | - | - | false |
| confirmButtonProps | 确认按钮属性 | <Link to="Button#Button API">ButtonProps</Link> | - | - | false |
| confirmButtonProps | 取消按钮属性 | <Link to="Button#Button API">ButtonProps</Link> | - | - | false |
| onConfirm | 点击确认事件 | (e)=>void | - | - | false |
| onCancel | 点击取消事件 | (e)=>void | - | - | false |
| placement | 指定弹出框的方位 | `auto`&#x007c;`auto-start`&#x007c;`auto-end`&#x007c;`top`&#x007c;`top-start`&#x007c;`top-end`&#x007c;`right`&#x007c;`right-start`&#x007c;`right-end`&#x007c;`bottom`&#x007c;`bottom-start`&#x007c;`bottom-end`&#x007c;`left`&#x007c;`left-start`&#x007c;`left-end` | - | `auto` | false |
| transitionClasses | 动画 class，参看<Link to='CSSTransition'>CSSTransition</Link> | boolean | - | `xl-popconfirm--zoom` | false |
| getPopupContainer | 指定弹出框的父容器 | string &#x007c; HTMLElement &#x007c; (() => string &#x007c; HTMLElement) | - | 默认为`body` | false |
| visible | 是否显示弹出框 | boolean | - | false | false |
| onVisibleChange | 弹出框状态改变时的回调 | (visible: boolean) => void | - | - | false |
| hideDelay | 延时关闭 | number | - | - | false |
| showDelay | 延时显示 | number | - | - | false |
| trigger | 触发方式 | `hover`&#x007c;`focus`&#x007c;`click`&#x007c;`contextMenu`&#x007c;`custom` | - | `click` | false |
| disablePopupEnter | 禁止鼠标进入弹出框 | boolean | - | - | false |
| forceRender | 强制弹出框立即挂载到 dom 中，即使弹出框还没有显示 | boolean | - | - | false |
| offset | 弹出框偏移 | `number` | - | `10` | false |
| style | 弹出框样式 | React.CSSProperties | - | - | false |
| className | 弹出框类名 | string | - | - | false |
| flip | 反转弹出框以避免遮挡 | boolean | HTMLElement | HTMLElement[] | (() =>)boolean | HTMLElement | HTMLElement[] | - | true | false |
| preventOverflow | 调整弹出框以避免遮挡 | boolean | HTMLElement | HTMLElement[] | (() =>)boolean | HTMLElement | HTMLElement[] | - | true | false |
| disableGpuAcceleration | 禁用 gpu 渲染 | boolean | - | - | false |
| disableTransformOrigin | 禁用默认的`transform-origin`行为 | boolean | - | - | false |
| disabled | 禁用 | boolean | - | - | false |
