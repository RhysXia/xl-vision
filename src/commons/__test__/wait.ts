import { act } from 'react-dom/test-utils'

export default (time: number) => {
  const wait = new Promise<void>(resolve => {
    // console.info(`wait time: ${time}`)
    setTimeout(() => {
      resolve()
    }, time)
  })

  return act(() => wait)
}
