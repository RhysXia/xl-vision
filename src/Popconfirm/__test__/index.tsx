import { mount, ReactWrapper } from 'enzyme'
import * as React from 'react'
import Popconfirm from '..'
import { act } from 'react-dom/test-utils'
import wait from '../../../test/wait'

describe('Popconfirm', () => {
  let wrapper: ReactWrapper

  beforeAll(() => {
    jest.useRealTimers()
  })

  afterEach(() => {
    wrapper.unmount()
  })

  it('render', () => {
    wrapper = mount(
      <Popconfirm className='popup' visible={true} message='message'>
        <button className='btn'>button</button>
      </Popconfirm>
    )

    expect(wrapper.render()).toMatchSnapshot()
  })

  it('测试onConfirm', async () => {
    const fn = jest.fn()
    wrapper = mount(
      <Popconfirm className='popup' visible={true} message='message' onConfirm={fn}>
        <button className='btn'>button</button>
      </Popconfirm>
    )

    await act(() => wait(50))

    const popup = document.querySelector('.popup') as HTMLElement

    expect(popup.style.display).toBe('')

    wrapper.find('button.xl-button--theme-primary').simulate('click')

    await act(() => wait(50))

    expect(fn.mock.calls.length).toBe(1)
    expect(popup.style.display).toBe('none')
  })

  it('测试onCancel', async () => {
    const fn = jest.fn()
    wrapper = mount(
      <Popconfirm className='popup' visible={true} message='message' onCancel={fn}>
        <button className='btn'>button</button>
      </Popconfirm>
    )

    await act(() => wait(50))

    const popup = document.querySelector('.popup') as HTMLElement

    expect(popup.style.display).toBe('')

    wrapper.find('button.xl-button--theme-default').simulate('click')

    await act(() => wait(50))

    expect(fn.mock.calls.length).toBe(1)
    expect(popup.style.display).toBe('none')
  })
})
