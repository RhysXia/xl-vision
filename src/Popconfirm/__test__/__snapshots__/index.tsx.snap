// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popconfirm render 1`] = `
Array [
  <button
    class="btn"
  >
    button
  </button>,
  <div
    data-popper-escaped=""
    data-popper-placement="top"
    data-popper-reference-hidden=""
    style="z-index: 1001; position: absolute; left: 0px; top: 0px; margin: 0px; padding-bottom: 10px; bottom: 0px; transform: translate(0px, 0px);"
  >
    <div
      class="xl-popconfirm popup"
      data-placement="top"
      style="transform-origin: 0px 100%;"
    >
      <div
        class="xl-popconfirm__arrow"
        data-popper-arrow=""
        style="position: absolute; left: 0px; transform: translate(0px, 0px);"
      />
      <div
        class="xl-popconfirm__inner"
      >
        <div
          class="xl-popconfirm__message"
        >
          <i
            class="xl-icon"
          >
            <svg
              style="height: 1em; width: 1em; fill: currentColor; stroke: currentColor;"
              viewBox="0 0 512 512"
            >
              <path
                d="M256 48C141.31 48 48 141.31 48 256s93.31 208 208 208 208-93.31 208-208S370.69 48 256 48zm0 319.91a20 20 0 1120-20 20 20 0 01-20 20zm21.72-201.15l-5.74 122a16 16 0 01-32 0l-5.74-121.94v-.05a21.74 21.74 0 1143.44 0z"
              />
            </svg>
          </i>
          <span
            class="xl-popconfirm__title"
          >
            message
          </span>
        </div>
        <div
          class="xl-popconfirm__footer"
        >
          <button
            class="xl-base-button xl-button xl-button--variant-contained xl-button--theme-default xl-button--size-small xl-button--elevation"
            tabindex="0"
          >
            <span
              class="xl-base-button__inner"
            >
              no
            </span>
            <span
              class="xl-ripple"
            />
          </button>
          <button
            class="xl-base-button xl-button xl-button--variant-contained xl-button--theme-primary xl-button--size-small xl-button--elevation"
            tabindex="0"
          >
            <span
              class="xl-base-button__inner"
            >
              yes
            </span>
            <span
              class="xl-ripple"
            />
          </button>
        </div>
      </div>
    </div>
  </div>,
]
`;
